vignettes/GenomicFiles.Rmd

```diff
@@ -0,0 +1,727 @@
+```
+
+File are supplied as a character vector or list of *File classes such as
+`BamFile`, `BigWigFile` etc.
+
+```{r class-bam-data}
+library(RNAseqData.HNRNPC.bam.chr14)
+fls <- RNAseqData.HNRNPC.bam.chr14_BAMFILES
+```
+
+The MAP function extracts and manipulates data subsets. Here we compute
+pileups for a given range and file.
+
+```{r quick_start-MAP}
+MAP <- function(range, file, ...) {
+    requireNamespace("Rsamtools")
+    Rsamtools::pileup(file, scanBamParam=Rsamtools::ScanBamParam(which=range))
+} 
+```
+
+`reduceByFile` sends each file to a worker where MAP is applied to each file /
+range combination. When `summarize=TRUE` the output is a `SummarizedExperiment`
+object.
+
+```{r quick_start-reduceByFile}
+se <- reduceByFile(gr, fls, MAP, summarize=TRUE)
+se 
+```
+
+Results are stored in the `assays` slot.
+
+```{r quick_start-assays}
+dim(assays(se)$data) ## ranges x files
+```
+
+`reduceByRange` sends each range to a worker and extracts the same range from
+all files. Adding a reducer to this example combines the pileups from each range
+across files.
+
+```{r quick_start-MAP-REDUCE-reduceByRange}
+library(GenomicAlignments)
+REDUCE <- function(mapped, ...) {
+    cmb = do.call(rbind, mapped)
+    xtabs(count ~ pos + nucleotide, cmb)
+}
+lst <- reduceByRange(gr, fls, MAP, REDUCE, iterate=FALSE)
+```
+
+The result is a list where each element is a summary table of counts for
+a single range across all 8 files.
+
+```{r quick_start-result}
+head(lst[[1]], 3)
+```
+
+# Overview of classes and functions
+
+## `GenomicFiles` class
+
+The `GenomicFiles` class is a matrix-like container where rows represent ranges
+of interest and columns represent files. The object can be subset on files and
+/ or ranges to perform different experimental runs. The class inherits from
+`RangedSummarizedExperiment` but does not (as of yet) make use of the
+`elementMetadata` and `assays` slots.
+
+```{r overview-GenomicFiles}
+GenomicFiles(gr, fls)
+```
+
+A `GenomicFiles` can be used as the `ranges` argument to the functions in this
+package. When `summarize=TRUE`, data from the common slots are transferred to
+the `SummarizedExperiment` result. NOTE: Results can only be put into a
+`SummarizedExperiment` when no reduction is performed because of the matching
+dimensions requirement (i.e., a REDUCE collapses the results in one dimension).
+
+## Functions
+
+Functions in `GenomicFiles` manipulate and combine data across or within files
+using the parallel infrastructure provided in `BiocParallel`. Files and ranges
+are sent to workers along with MAP and REDUCE functions. The MAP extracts and/or
+manipulates data and REDUCE consolidates the results from MAP. Both MAP and
+REDUCE are executed in the distributed step and therefore reduction occurs on
+data from the same worker, not across workers. The chart in Figure
+[1](reduceByRange_flow.png) represents the division of labor in `reduceByRange`
```

Comment:
Rather than link directly to the image, it would be better to link to the place of image in the document. I think it is okay to use an anchor with id before the figure `<a id="fig1"></a>` then I think you can do `[1](#fig1)`